File: 9/a.nim
import strutils, sets

let
  file: File = open("input.txt")
  input: string = file.readAll()
  lines: seq[string] = input.splitLines

var visited: HashSet[(int, int)]

var
  H = (0, 0)
  T = (0, 0)

template x(pair: (int, int)): int = pair[0]

template y(pair: (int, int)): int = pair[1]

for line in lines:
  let
    items: seq[string] = line.splitWhitespace()
    letter: char = items[0][0]
    number: int = parseInt items[1]

  for i in 0..<number:
    let
      dx = H.x - T.x
      dy = H.y - T.y

    case letter:
      of 'R':
        if 1 == dx: T = H
        inc H.x
      of 'L':
        if -1 == dx: T = H
        dec H.x
      of 'U':
        if 1 == dy: T = H
        inc H.y
      of 'D':
        if -1 == dy: T = H
        dec H.y
      else: discard
    visited.incl(T)

echo visited.len

File: 9/a.cpp
#include "utils.h"

long long solve(vector<string> &lines) {
  set<pair<int, int>> visited;

  pair<int /*y*/, int /*x*/> H{0, 0}, T{0, 0};
  for (string &line : lines) {
    stringstream ss(line);
    char letter;
    int number;
    ss >> letter >> number;
    for (int i = 0; i < number; ++i) {
      if ('R' == letter) {
        if (1 == H.second - T.second) {
          T = H;
        }
        ++H.second;
      } else if ('L' == letter) {
        if (-1 == H.second - T.second) {
          T = H;
        }
        --H.second;
      } else if ('U' == letter) {
        if (1 == H.first - T.first) {
          T = H;
        }
        ++H.first;
      } else if ('D' == letter) {
        if (-1 == H.first - T.first) {
          T = H;
        }
        --H.first;
      }
      visited.insert(T);
    }
  }

  return visited.size();
}